File: packages/protocol/src/accounts.ts
/**
 * The exact bytes an organisation's policy-signing key covers.
 *
 * A signature is over bytes, never over JSON. Two encoders that agree on a document can still
 * disagree on its text, and a verifier that re-serialises what it received is checking its own
 * encoder rather than the issuer's statement. So each payload is built here, once, from the JSON
 * representation the managed service carries, and both the issuer and a verifier cover identical
 * bytes. The Rust host builds the same payloads from `kr_protocol::account`; `fixtures/accounts`
 * publishes the bytes both languages must produce.
 *
 * Each schema is closed. A payload with a field nobody agreed on, a counter that is not an exact
 * unsigned integer, or a right outside the vocabulary is refused rather than narrowed to the part
 * this version understands: signing what is left of a document after discarding what a verifier
 * did not recognise is exactly the strip-and-verify behaviour section 23 forbids.
 */

import { encodeCanonical } from './cbor/encode.js'
import {
  krArray,
  krBytes,
  krInt,
  krMap,
  krNull,
  krText,
  type CanonicalValue
} from './cbor/value.js'
import { base64UrlToBytes, jsonToU64, jsonToUuid } from './json.js'
import type {
  ActionRight,
  MembershipLeasePayload,
  PolicyAuthorityHeadPayload,
  PolicyAuthorityLinkPayload
} from './generated/protocol.js'

/** The domain a membership lease signature covers. */
export const MEMBERSHIP_LEASE_DOMAIN = 'kr-membership-lease/1'

/** The domain one policy-signing authority link covers. */
export const POLICY_AUTHORITY_DOMAIN = 'kr-policy-authority/1'

/** The domain the statement of the current revision covers. */
export const POLICY_AUTHORITY_HEAD_DOMAIN = 'kr-policy-authority-head/1'

/** The longest a membership lease may last, in milliseconds. */
export const MEMBERSHIP_LEASE_MAX_LIFETIME_MS = 15 * 60 * 1000

/** How often a holder asks for the next lease, in milliseconds. */
export const MEMBERSHIP_LEASE_REFRESH_INTERVAL_MS = 5 * 60 * 1000

/** The longest a head statement may last, in milliseconds. */
export const POLICY_AUTHORITY_HEAD_MAX_LIFETIME_MS = MEMBERSHIP_LEASE_MAX_LIFETIME_MS

/** Raw bytes in an Ed25519 public key. */
export const ED25519_PUBLIC_KEY_BYTES = 32

/** Raw bytes in an Ed25519 signature. */
export const ED25519_SIGNATURE_BYTES = 64

/** An organisation role, from the least to the most authority. */
export type TeamRole = MembershipLeasePayload['role']

/** The roles, from the least to the most authority. */
export const TEAM_ROLES: readonly TeamRole[] = ['viewer', 'reviewer', 'controller', 'owner']

/**
 * The most each role may ever carry.
 *
 * A role is a label for a ceiling; a host authorises from the rights a lease names, never from the
 * label. Answering a question is inside a viewer's and a reviewer's ceiling because an organisation
 * may invite one to answer, which is not the same as carrying it by default.
 */
export const ROLE_MAXIMUM_GRANTS: Readonly<Record<TeamRole, readonly ActionRight[]>> = {
  viewer: ['question.respond', 'session.view'],
  reviewer: ['files.read', 'question.respond', 'session.view'],
  controller: ['files.read', 'question.respond', 'session.view', 'terminal.input'],
  owner: [
    'files.read',
    'question.respond',
    'session.close',
    'session.share',
    'session.view',
    'terminal.input'
  ]
}

/** Raised when a payload does not match the closed schema its signature covers. */
export class AccountSchemaError extends Error {
  constructor (message: string) {
    super(message)
    this.name = 'AccountSchemaError'
  }
}

function refuse (message: string): never {
  throw new AccountSchemaError(message)
}

/** The fields of a closed schema, with nothing else beside them. */
function closed (what: string, value: unknown, fields: readonly string[]): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    refuse(`${what} is an object`)
  }
  const record = value as Record<string, unknown>
  for (const key of Object.keys(record)) {
    if (!fields.includes(key)) {
      refuse(`${what} has no field ${JSON.stringify(key)}`)
    }
  }
  for (const field of fields) {
    if (!Object.hasOwn(record, field)) {
      refuse(`${what} is missing ${JSON.stringify(field)}`)
    }
  }
  return record
}

/** An identifier the protocol carries as 16 bytes, from its hyphenated text. */
function identifier (what: string, value: unknown): CanonicalValue {
  if (typeof value !== 'string') {
    refuse(`${what} is a hyphenated identifier`)
  }
  try {
    return krBytes(jsonToUuid(value))
  } catch (error) {
    refuse(`${what} is a hyphenated identifier: ${error instanceof Error ? error.message : 'invalid'}`)
  }
}

/** Longest an opaque identifier may be, in bytes, as `kr_protocol::ids` bounds one. */
const MAX_OPAQUE_ID_BYTES = 256

/**
 * An identifier the service mints and the protocol carries verbatim.
 *
 * A host never resolves it: it names a person to an organisation. It is bounded
 * and free of control characters, so a peer cannot force an unbounded
 * allocation, or a line break, through an identifier field.
 */
function opaqueIdentifier (what: string, value: unknown): CanonicalValue {
  if (typeof value !== 'string' || value === '') {
    refuse(`${what} is a non-empty identifier`)
  }
  if (new TextEncoder().encode(value).length > MAX_OPAQUE_ID_BYTES) {
    refuse(`${what} is at most ${String(MAX_OPAQUE_ID_BYTES)} bytes`)
  }
  // eslint-disable-next-line no-control-regex
  if (/[\u0000-\u001f\u007f]/.test(value)) {
    refuse(`${what} carries no control characters`)
  }
  return krText(value)
}

/**
 * A counter that travelled as a decimal string, as a canonical integer.
 *
 * JSON carries an unsigned 64-bit value as text because a JSON number cannot hold one exactly. A
 * number is refused here rather than converted, even one a reader could convert exactly: `2`, `2.0`
 * and `2e0` are one JavaScript value and three documents, and the host refuses all three. Two
 * producers that disagreed about which of them to send would otherwise sign identical bytes from
 * documents a strict reader treats as different.
 */
function counter (what: string, value: unknown): CanonicalValue {
  if (typeof value !== 'string') {
    refuse(`${what} is an unsigned counter written as a decimal string`)
  }
  try {
    return krInt(jsonToU64(value))
  } catch (error) {
    refuse(`${what} is an unsigned counter: ${error instanceof Error ? error.message : 'invalid'}`)
  }
}

/** A fixed-width scalar read from its base64url text. */
function scalar (what: string, value: unknown, width: number): Uint8Array {
  if (typeof value !== 'string') {
    refuse(`${what} is unpadded base64url`)
  }
  let bytes: Uint8Array
  try {
    bytes = base64UrlToBytes(value)
  } catch (error) {
    refuse(`${what} is unpadded base64url: ${error instanceof Error ? error.message : 'invalid'}`)
  }
  if (bytes.length !== width) {
    refuse(`${what} is ${String(width)} bytes, not ${String(bytes.length)}`)
  }
  return bytes
}

/** An Ed25519 public key read from the base64url text a response carries. */
export function publicKeyBytes (value: unknown): Uint8Array {
  return scalar('an Ed25519 public key', value, ED25519_PUBLIC_KEY_BYTES)
}

/** An Ed25519 signature read from the base64url text a response carries. */
export function signatureBytes (value: unknown): Uint8Array {
  return scalar('an Ed25519 signature', value, ED25519_SIGNATURE_BYTES)
}

function signingInput (domain: string, payload: CanonicalValue): Uint8Array {
  return encodeCanonical(krArray([krText(domain), payload]))
}

const LEASE_FIELDS = [
  'account_id',
  'expires_at_ms',
  'issued_at_ms',
  'key_revision',
  'maximum_grants',
  'organisation_id',
  'role'
] as const

/**
 * The closed action-right vocabulary, in the order the protocol declares it.
 *
 * A lease names rights from this vocabulary and no others, so a host intersects what a lease says
 * with the grants it already understands rather than with a second set of names.
 */
export const ACTION_RIGHTS: readonly ActionRight[] = [
  'session.view',
  'terminal.input',
  'terminal.geometry',
  'terminal.geometry.transfer',
  'terminal.palette',
  'agent.prompt',
  'agent.cancel',
  'agent.approval.respond',
  'question.respond',
  'files.read',
  'files.upload',
  'files.apply_diff',
  'project.create',
  'workspace.manage',
  'changeset.create',
  'session.create',
  'session.rename',
  'session.close',
  'session.share',
  'automation.manage',
  'host.manage'
]

/**
 * The rights a set carries, in the one order the wire form admits.
 *
 * Rights are ordered by their wire string and carry no duplicates, so a set encodes to the same
 * bytes whatever order an issuer holds it in. A set that is not already in that order is refused
 * rather than sorted: a verifier checks the bytes it received, so re-ordering them here would cover
 * bytes the issuer never sent. An issuer calls {@link canonicalRights} once, before it signs.
 */
function rights (what: string, value: unknown): CanonicalValue {
  if (!Array.isArray(value)) {
    refuse(`${what} is an array`)
  }
  const members: ActionRight[] = []
  for (const member of value as unknown[]) {
    if (typeof member !== 'string' || !(ACTION_RIGHTS as readonly string[]).includes(member)) {
      refuse(`${JSON.stringify(member)} is not an action right`)
    }
    members.push(member as ActionRight)
  }
  for (let index = 1; index < members.length; index += 1) {
    if ((members[index - 1] as string) >= (members[index] as string)) {
      refuse(`${what} is in ascending order without duplicates`)
    }
  }
  return krArray(members.map((right) => krText(right)))
}

/**
 * Sorts and de-duplicates a set of rights into the order a lease states them in.
 *
 * An issuer holding a role's ceiling calls this once and stores the result; the lease it signs
 * then carries the same bytes as a lease any other issuer of the same set would sign.
 */
export function canonicalRights (value: readonly ActionRight[]): ActionRight[] {
  return [...new Set(value)].sort() as ActionRight[]
}

/**
 * The bytes a membership lease is signed over.
 *
 * `CBOR(["kr-membership-lease/1", the payload as a canonical map])`.
 *
 * Everything a standalone authorisation object must cover is here: the type through the domain,
 * the issuing organisation and the key revision that signed it, the account it names, the authority
 * it carries, and when it ends.
 */
export function membershipLeaseSigningInput (payload: MembershipLeasePayload): Uint8Array {
  const record = closed('a membership lease', payload, LEASE_FIELDS)
  const role = record['role']
  if (typeof role !== 'string' || !(TEAM_ROLES as readonly string[]).includes(role)) {
    refuse(`${JSON.stringify(role)} is not a team role`)
  }
  return signingInput(
    MEMBERSHIP_LEASE_DOMAIN,
    krMap([
      ['account_id', opaqueIdentifier('an account identifier', record['account_id'])],
      ['expires_at_ms', counter('a lease expiry', record['expires_at_ms'])],
      ['issued_at_ms', counter('a lease issue time', record['issued_at_ms'])],
      ['key_revision', counter('a key revision', record['key_revision'])],
      ['maximum_grants', rights('a grant ceiling', record['maximum_grants'])],
      ['organisation_id', identifier('an organisation identifier', record['organisation_id'])],
      ['role', krText(role)]
    ])
  )
}

const LINK_FIELDS = [
  'key_revision',
  'not_before_ms',
  'organisation_id',
  'previous_key_revision',
  'public_key'
] as const

/**
 * The bytes one authority-chain link is signed over.
 *
 * `CBOR(["kr-policy-authority/1", the payload as a canonical map])`.
 *
 * The first revision signs its own link and names no predecessor, which is what a host pins. Every
 * later revision is signed by the key of the revision it names, so a host that pinned the first can
 * walk forward to the current key without being told which key to trust, and a link cannot be
 * re-parented under a revision it was not issued against.
 */
export function policyAuthorityLinkSigningInput (payload: PolicyAuthorityLinkPayload): Uint8Array {
  const record = closed('an authority link', payload, LINK_FIELDS)
  const previous = record['previous_key_revision']
  return signingInput(
    POLICY_AUTHORITY_DOMAIN,
    krMap([
      ['key_revision', counter('a key revision', record['key_revision'])],
      ['not_before_ms', counter('an activation time', record['not_before_ms'])],
      ['organisation_id', identifier('an organisation identifier', record['organisation_id'])],
      [
        'previous_key_revision',
        previous === null ? krNull() : counter('a predecessor revision', previous)
      ],
      ['public_key', krBytes(publicKeyBytes(record['public_key']))]
    ])
  )
}

const HEAD_FIELDS = ['expires_at_ms', 'issued_at_ms', 'key_revision', 'organisation_id'] as const

/**
 * The bytes the statement naming the current revision is signed over.
 *
 * `CBOR(["kr-policy-authority-head/1", the payload as a canonical map])`.
 *
 * Any prefix of a chain verifies on its own, so a host handed the first few links cannot tell from
 * the links alone that a later revision exists. This statement is what tells it: it is signed by
 * the key that signs leases now, and it expires, so a captured statement stops being usable and a
 * host that has accepted a later revision refuses one naming an earlier revision. A retained
 * private key of a retired revision can still sign statements naming that revision, which is why
 * rotation destroys it.
 */
export function policyAuthorityHeadSigningInput (payload: PolicyAuthorityHeadPayload): Uint8Array {
  const record = closed('an authority head', payload, HEAD_FIELDS)
  return signingInput(
    POLICY_AUTHORITY_HEAD_DOMAIN,
    krMap([
      ['expires_at_ms', counter('a head expiry', record['expires_at_ms'])],
      ['issued_at_ms', counter('a head issue time', record['issued_at_ms'])],
      ['key_revision', counter('a key revision', record['key_revision'])],
      ['organisation_id', identifier('an organisation identifier', record['organisation_id'])]
    ])
  )
}

File: crates/kr-protocol/tests/accounts.rs
//! Cross-language conformance for the account vectors under `fixtures/accounts/`.
//!
//! The TypeScript package builds the same payloads from their JSON representation and asserts the
//! same bytes, so the managed service and a host sign and verify over identical input.

use std::path::PathBuf;
use std::str::FromStr;

use kr_cbor::{CanonicalMap, CanonicalValue, Limits, decode, encode};
use kr_protocol::account::{
    ChainError, MEMBERSHIP_LEASE_DOMAIN, MEMBERSHIP_LEASE_MAX_LIFETIME_MS, MembershipLease,
    MembershipLeasePayload, POLICY_AUTHORITY_DOMAIN, POLICY_AUTHORITY_HEAD_DOMAIN, PolicyAuthority,
    PolicyAuthorityHead, PolicyAuthorityHeadPayload, PolicyAuthorityLink,
    PolicyAuthorityLinkPayload, TeamRole,
};
use kr_protocol::ids::{AccountId, OrganisationId, PolicyKeyRevision};
use kr_protocol::rights::ActionRight;
use kr_protocol::scalars::{AuthorisationKey, Nullable, Signature64, TimestampMs, Uuid};
use serde_json::Value as Json;

fn load() -> Json {
    let path =
        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../fixtures/accounts/leases.json");
    let text = std::fs::read_to_string(&path)
        .unwrap_or_else(|error| panic!("read {}: {error}", path.display()));
    serde_json::from_str(&text).unwrap_or_else(|error| panic!("parse {}: {error}", path.display()))
}

fn case<'a>(document: &'a Json, id: &str) -> &'a Json {
    document["cases"]
        .as_array()
        .expect("cases")
        .iter()
        .find(|case| case["id"] == id)
        .unwrap_or_else(|| panic!("case {id}"))
}

/// Parses the fixture value grammar into a canonical value.
fn parse_value(description: &Json) -> CanonicalValue {
    let object = description.as_object().expect("value description object");
    assert_eq!(object.len(), 1, "a value description has exactly one key");
    let (kind, payload) = object.iter().next().expect("one entry");
    match kind.as_str() {
        "int" => CanonicalValue::integer(
            payload
                .as_str()
                .expect("decimal string")
                .parse::<i128>()
                .expect("decimal integer"),
        )
        .expect("inside the 64-bit argument range"),
        "bytes" => {
            CanonicalValue::Bytes(hex::decode(payload.as_str().expect("hex")).expect("valid hex"))
        }
        "text" => CanonicalValue::text(payload.as_str().expect("string")),
        "bool" => CanonicalValue::Bool(payload.as_bool().expect("boolean")),
        "null" => CanonicalValue::Null,
        "array" => CanonicalValue::Array(
            payload
                .as_array()
                .expect("array")
                .iter()
                .map(parse_value)
                .collect(),
        ),
        "map" => {
            let mut map = CanonicalMap::new();
            for entry in payload.as_array().expect("array of pairs") {
                let pair = entry.as_array().expect("pair");
                map.insert(
                    pair[0].as_str().expect("text key").to_owned(),
                    parse_value(&pair[1]),
                )
                .expect("no duplicate keys");
            }
            CanonicalValue::Map(map)
        }
        other => panic!("unknown value kind {other}"),
    }
}

fn uuid(text: &str) -> Uuid {
    Uuid::from_str(text).expect("valid uuid")
}

fn organisation() -> OrganisationId {
    OrganisationId::new(uuid("8f14e45f-ea1e-4b9e-9f3a-0a3a9b0d2f61"))
}

fn account() -> AccountId {
    AccountId::new("3c9f2b7a-5d18-4a62-9c07-1f5b8e2d4a90").expect("account identifier")
}

fn owner_lease() -> MembershipLease {
    MembershipLease {
        payload: MembershipLeasePayload {
            organisation_id: organisation(),
            account_id: account(),
            role: TeamRole::Owner,
            maximum_grants: TeamRole::Owner.maximum_grants(),
            issued_at_ms: TimestampMs::new(1_767_225_600_000),
            expires_at_ms: TimestampMs::new(1_767_225_600_000 + MEMBERSHIP_LEASE_MAX_LIFETIME_MS),
            key_revision: PolicyKeyRevision::new(2),
        },
        signature: Signature64::from_bytes([0x5a; 64]),
    }
}

fn narrowed_lease() -> MembershipLease {
    MembershipLease {
        payload: MembershipLeasePayload {
            organisation_id: organisation(),
            account_id: account(),
            role: TeamRole::Reviewer,
            maximum_grants: [ActionRight::SessionView, ActionRight::FilesRead]
                .into_iter()
                .collect(),
            issued_at_ms: TimestampMs::new(1_767_225_600_000),
            expires_at_ms: TimestampMs::new(1_767_225_600_000 + 5 * 60 * 1000),
            key_revision: PolicyKeyRevision::new(2),
        },
        signature: Signature64::from_bytes([0x6b; 64]),
    }
}

fn first_link() -> PolicyAuthorityLink {
    PolicyAuthorityLink {
        payload: PolicyAuthorityLinkPayload {
            organisation_id: organisation(),
            key_revision: PolicyKeyRevision::new(1),
            previous_key_revision: Nullable::null(),
            public_key: AuthorisationKey::from_bytes([0x11; 32]),
            not_before_ms: TimestampMs::new(1_767_139_200_000),
        },
        signature: Signature64::from_bytes([0x21; 64]),
    }
}

fn rotation_link() -> PolicyAuthorityLink {
    PolicyAuthorityLink {
        payload: PolicyAuthorityLinkPayload {
            organisation_id: organisation(),
            key_revision: PolicyKeyRevision::new(2),
            previous_key_revision: Nullable::some(PolicyKeyRevision::new(1)),
            public_key: AuthorisationKey::from_bytes([0x22; 32]),
            not_before_ms: TimestampMs::new(1_767_182_400_000),
        },
        signature: Signature64::from_bytes([0x32; 64]),
    }
}

fn head() -> PolicyAuthorityHead {
    PolicyAuthorityHead {
        payload: PolicyAuthorityHeadPayload {
            organisation_id: organisation(),
            key_revision: PolicyKeyRevision::new(2),
            issued_at_ms: TimestampMs::new(1_767_225_600_000),
            expires_at_ms: TimestampMs::new(1_767_225_600_000 + 15 * 60 * 1000),
        },
        signature: Signature64::from_bytes([0x43; 64]),
    }
}

fn published_authority() -> PolicyAuthority {
    PolicyAuthority {
        organisation_id: organisation(),
        chain: vec![first_link(), rotation_link()],
        head: head(),
    }
}

/// Asserts that `signing_input` matches the fixture bytes, its digest and its value description,
/// and that the record's JSON representation round-trips through the fixture document.
fn assert_vector<T>(document: &Json, id: &str, record: &T, signing_input: &[u8])
where
    T: serde::Serialize + serde::de::DeserializeOwned + PartialEq + std::fmt::Debug,
{
    let case = case(document, id);
    let expected_hex = case["cbor_hex"].as_str().expect("cbor_hex");

    assert_eq!(hex::encode(signing_input), expected_hex, "{id}: bytes");
    assert_eq!(
        hex::encode(kr_cbor::sha256(signing_input)),
        case["sha256"].as_str().expect("sha256"),
        "{id}: digest"
    );

    let described = parse_value(&case["value"]);
    assert_eq!(
        hex::encode(encode(&described)),
        expected_hex,
        "{id}: the fixture description and its bytes disagree"
    );
    assert_eq!(
        decode(&hex::decode(expected_hex).expect("hex"), &Limits::DEFAULT).expect("decode"),
        described,
        "{id}: decoded value"
    );

    let json = &case["json"];
    let parsed: T = serde_json::from_value(json.clone())
        .unwrap_or_else(|error| panic!("{id}: the JSON representation does not parse: {error}"));
    assert_eq!(&parsed, record, "{id}: the JSON representation");
    assert_eq!(
        &serde_json::to_value(record).expect("serialise"),
        json,
        "{id}: the JSON the record produces"
    );
}

#[test]
fn membership_lease_vectors_match() {
    let document = load();
    let lease = owner_lease();
    assert_vector(
        &document,
        "membership_lease",
        &lease,
        &lease.payload.signing_input().expect("signing input"),
    );

    let narrowed = narrowed_lease();
    assert_vector(
        &document,
        "membership_lease_narrowed",
        &narrowed,
        &narrowed.payload.signing_input().expect("signing input"),
    );
}

#[test]
fn policy_authority_vectors_match() {
    let document = load();

    let first = first_link();
    assert_vector(
        &document,
        "policy_authority_first_revision",
        &first,
        &first.payload.signing_input().expect("signing input"),
    );

    let rotation = rotation_link();
    assert_vector(
        &document,
        "policy_authority_rotation",
        &rotation,
        &rotation.payload.signing_input().expect("signing input"),
    );

    let head = head();
    assert_vector(
        &document,
        "policy_authority_head",
        &head,
        &head.payload.signing_input().expect("signing input"),
    );
}

#[test]
fn every_domain_is_published_with_its_vector() {
    let document = load();
    for (id, domain) in [
        ("membership_lease", MEMBERSHIP_LEASE_DOMAIN),
        ("membership_lease_narrowed", MEMBERSHIP_LEASE_DOMAIN),
        ("policy_authority_first_revision", POLICY_AUTHORITY_DOMAIN),
        ("policy_authority_rotation", POLICY_AUTHORITY_DOMAIN),
        ("policy_authority_head", POLICY_AUTHORITY_HEAD_DOMAIN),
    ] {
        assert_eq!(case(&document, id)["domain"], domain, "{id}: domain");
    }
}

#[test]
fn a_domain_separates_one_statement_from_another() {
    // The head and a lease share their organisation, revision and times. Only the domain and the
    // field set differ, which is what stops one signature being read as the other.
    let lease = owner_lease().payload.signing_input().expect("lease");
    let head = head().payload.signing_input().expect("head");
    assert_ne!(lease, head);
    assert!(lease.starts_with(&[0x82]), "an array of two elements");
    assert!(head.starts_with(&[0x82]), "an array of two elements");
}

#[test]
fn a_lease_states_a_ceiling_inside_its_role() {
    let lease = owner_lease();
    assert!(lease.payload.grants_within_role());
    assert!(lease.payload.lifetime_within_maximum());
    assert!(narrowed_lease().payload.grants_within_role());

    let mut widened = narrowed_lease();
    widened.payload.maximum_grants = TeamRole::Owner.maximum_grants();
    assert!(
        !widened.payload.grants_within_role(),
        "a reviewer cannot carry an owner's ceiling"
    );

    let mut long = owner_lease();
    long.payload.expires_at_ms =
        TimestampMs::new(long.payload.issued_at_ms.get() + MEMBERSHIP_LEASE_MAX_LIFETIME_MS + 1);
    assert!(!long.payload.lifetime_within_maximum());
}

#[test]
fn a_lease_is_usable_only_inside_its_window() {
    let lease = owner_lease();
    let issued = lease.payload.issued_at_ms.get();
    let expires = lease.payload.expires_at_ms.get();
    assert!(!lease.payload.is_valid_at(issued - 1));
    assert!(lease.payload.is_valid_at(issued));
    assert!(lease.payload.is_valid_at(expires - 1));
    assert!(!lease.payload.is_valid_at(expires));
}

#[test]
fn a_published_chain_is_checked_before_any_signature() {
    let authority = published_authority();
    assert_eq!(authority.check_structure(), Ok(()));
    assert_eq!(authority.current_revision(), PolicyKeyRevision::new(2));
    assert_eq!(
        authority.link(PolicyKeyRevision::new(1)),
        Some(&first_link())
    );

    let mut empty = published_authority();
    empty.chain.clear();
    assert_eq!(empty.check_structure(), Err(ChainError::Empty));

    let mut reparented = published_authority();
    reparented.chain[1].payload.previous_key_revision = Nullable::some(PolicyKeyRevision::new(2));
    assert_eq!(
        reparented.check_structure(),
        Err(ChainError::BrokenSuccession { index: 1 })
    );

    let mut self_signed_successor = published_authority();
    self_signed_successor.chain[1].payload.previous_key_revision = Nullable::null();
    assert_eq!(
        self_signed_successor.check_structure(),
        Err(ChainError::BrokenSuccession { index: 1 })
    );

    let mut gap = published_authority();
    gap.chain[1].payload.key_revision = PolicyKeyRevision::new(3);
    assert_eq!(
        gap.check_structure(),
        Err(ChainError::UnorderedRevisions { index: 1 })
    );

    let mut backdated = published_authority();
    backdated.chain[1].payload.not_before_ms = TimestampMs::new(1);
    assert_eq!(
        backdated.check_structure(),
        Err(ChainError::UnorderedActivation { index: 1 })
    );

    let mut truncated = published_authority();
    truncated.chain.pop();
    assert_eq!(
        truncated.check_structure(),
        Err(ChainError::HeadRevisionMismatch)
    );

    let mut foreign = published_authority();
    foreign.chain[1].payload.organisation_id =
        OrganisationId::new(uuid("00000000-0000-4000-8000-000000000000"));
    assert_eq!(
        foreign.check_structure(),
        Err(ChainError::ForeignOrganisation { index: 1 })
    );

    let mut foreign_head = published_authority();
    foreign_head.head.payload.organisation_id =
        OrganisationId::new(uuid("00000000-0000-4000-8000-000000000000"));
    assert_eq!(
        foreign_head.check_structure(),
        Err(ChainError::HeadForeignOrganisation)
    );

    let mut early_head = published_authority();
    early_head.head.payload.issued_at_ms =
        TimestampMs::new(early_head.chain[1].payload.not_before_ms.get() - 1);
    assert_eq!(
        early_head.check_structure(),
        Err(ChainError::HeadBeforeActivation),
        "a head cannot say a revision signs now before that revision took over"
    );

    let mut exactly_at_activation = published_authority();
    let activation = exactly_at_activation.chain[1].payload.not_before_ms.get();
    exactly_at_activation.head.payload.issued_at_ms = TimestampMs::new(activation);
    exactly_at_activation.head.payload.expires_at_ms =
        TimestampMs::new(activation + 15 * 60 * 1000);
    assert_eq!(
        exactly_at_activation.check_structure(),
        Ok(()),
        "a head issued at the moment of activation is admissible"
    );

    let mut long_head = published_authority();
    long_head.head.payload.expires_at_ms =
        TimestampMs::new(long_head.head.payload.issued_at_ms.get() + 16 * 60 * 1000);
    assert_eq!(long_head.check_structure(), Err(ChainError::HeadLifetime));

    let mut premature = published_authority();
    premature.chain[0].payload.previous_key_revision = Nullable::some(PolicyKeyRevision::new(0));
    assert_eq!(
        premature.check_structure(),
        Err(ChainError::FirstLinkHasPredecessor)
    );
}

#[test]
fn a_chain_authenticates_forward_from_the_revision_a_host_pinned() {
    let authority = published_authority();

    // A host that pinned the first revision reaches both: the pin, and the
    // revision the pin's key signed the link for.
    let from_first = authority
        .authenticated_from(PolicyKeyRevision::new(1))
        .expect("the chain carries revision 1");
    assert_eq!(from_first.len(), 2);

    // A host that pinned the second reaches the second and anything after it,
    // and not the first: nothing it has authenticates a revision from before its
    // pin, so a lease signed by that revision is refused rather than checked
    // against a key it has no reason to trust.
    let from_second = authority
        .authenticated_from(PolicyKeyRevision::new(2))
        .expect("the chain carries revision 2");
    assert_eq!(from_second.len(), 1);
    assert_eq!(from_second[0], rotation_link());
    assert!(
        !from_second
            .iter()
            .any(|link| link.payload.key_revision == PolicyKeyRevision::new(1))
    );

    assert!(
        authority
            .authenticated_from(PolicyKeyRevision::new(3))
            .is_none()
    );
}

#[test]
fn a_closed_schema_refuses_a_field_nobody_agreed_on() {
    let mut json = serde_json::to_value(owner_lease()).expect("serialise");
    json["payload"]["scope"] = Json::String("everything".to_owned());
    let parsed: Result<MembershipLease, _> = serde_json::from_value(json);
    assert!(parsed.is_err(), "an unknown field is refused");

    let mut unsorted = serde_json::to_value(owner_lease()).expect("serialise");
    unsorted["payload"]["maximum_grants"] = Json::Array(vec![
        Json::String("session.view".to_owned()),
        Json::String("files.read".to_owned()),
    ]);
    let parsed: Result<MembershipLease, _> = serde_json::from_value(unsorted);
    assert!(
        parsed.is_err(),
        "a ceiling that arrived out of order is refused rather than sorted, because a verifier \
         checks the bytes it received"
    );

    let mut unknown_grant = serde_json::to_value(owner_lease()).expect("serialise");
    unknown_grant["payload"]["maximum_grants"] =
        Json::Array(vec![Json::String("everything".to_owned())]);
    let parsed: Result<MembershipLease, _> = serde_json::from_value(unknown_grant);
    assert!(parsed.is_err(), "a right outside the vocabulary is refused");
}
